libass/ass_parse.c

```diff
@@ -128,21 +128,53 @@ void update_font(ASS_Renderer *render_priv)
     render_priv->state.font = ass_font_new(render_priv, &desc);
 }
 
+static double calc_anim(double new, double old, double pwr)
+{
+   return (1 - pwr) * old + new * pwr;
+}
+
+static uint32_t calc_anim_uint32(uint32_t new, uint32_t old, double pwr)
+{
+    double ret = calc_anim(new, old, pwr);
+
+    // Avoid UB on out-of-range values; match x86 behavior
+    if (isnan(ret) || ret < 0 || ret > UINT32_MAX)
+        return INT32_MIN;
```

Comment:
Is this correct for the `uint32_t` version?

Why are you using an `uint32_t` variant for alpha to begin with? As far as I see, for alpha VSFilter casts the double to `BYTE`, which is `unsigned char`. Wouldn’t that produce something different? From a quick & dirty test on Intel macOS (not sure how representative that is), I seem to get 0 from all kinds of unrepresentable values when casting `double` to `unsigned char`. For reference, I see `(int)CalcAnimation` and `(BYTE)CalcAnimation` in VSFilter, and no other variants.

Reply:
> Is this correct for the `uint32_t` version?

Oh, hmmm, I was looking at the documentation for `cvttsd2si`, which returns `0x80000000` or `0x8000000000000000` on error… but it looks like compilers actually explicitly handle this when converting to an integer type and give 0 instead? Huh. It doesn't affect the color case regardless, since the high bit will get truncated away, but sure I'll switch this to return 0.

Looking closer, apparently negative (but ≥INT32_MIN) values actually convert to their signed integer representations and then cast to unsigned from there, so I'll correct that (don't think this can happen in the color case either but might as well be thorough).

Fun fact I just found while investigating deeper: signed int behavior seems to be the same regardless, but converting double to unsigned int can give different results when building for 32- vs 64-bit! I think we should probably go with the 32-bit behavior.

So, my implementation of `calc_anim_int32` is correct, but `calc_anim_uint32` should:
- Return 0 for values above `UINT32_T`
- Return 0 for NAN
- Return `(uint32_t)INT_MIN` for values below `INT_MIN`
- Return `(uint32_t)(int32_t)ret` for values in `[INT_MIN, 0)`
- Return `(uint32_t)ret` for values in `[0, UINT32_MAX]`

As far as I can tell, casting from double to char (either signedness) always goes via converting to a signed int, then just extracting the low byte. So yes, we should be using the signed version there.

…Which also means the unsigned version isn't actually needed at all, is it? So I can just drop all that complexity.